### PR TITLE
[unittests] OperationGradTest: add a unittest for conv operator.

### PR DESCRIPTION
**Pull request by d1jang**
*Description*:

This PR adds a unittest for conv to OperationGradTest.

*Testing*:

Ran OperationGradTest

*Documentation*:

This is a very basic test for "conv" operator and more tests will follow to enhance its coverage.

Partially fixing #2513

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

File: tests/unittests/OperatorGradTest.cpp [hunk 1]
```diff
@@ -105,6 +105,57 @@ TEST_P(OperatorGradTest, concat) {
...
+  // Note that filter is trainable, but probably it's okay since we run the
+  // graph only once.
+  bindings_.get(llvm::cast<Placeholder>(O->getFilter()))->getHandle() = {0, 1,
+                                                                         2, 3};
+  bindings_.get(llvm::cast<Placeholder>(O->getBias()))->getHandle() = {1};
+
+  VariableGradientsList varGrads = computeVarGrads(O);
+
+  // (Note that T stands for the filter).
+  // O_0_0 = X_0_0*T_0_0 + X_0_1*T_0_1 + X_1_0*T_1_0 + X_1_1*T_1_1 + b = 20
```
Comment (SplitInfinity) [inline tests/unittests/OperatorGradTest.cpp]: Consider writing these equations as

```
O_{0,0} = X_{0,0}*T_{0,0} + X_{0,1}*T_{0,1} + X_{1,0}*T_{1,0} + X_{1,1}*T_{1,1} + b = 20
```

I think they will be clearer this way.

File: tests/unittests/OperatorGradTest.cpp [hunk 2]
```diff
@@ -105,6 +105,57 @@ TEST_P(OperatorGradTest, concat) {
...
+  // O_0_1 = X_0_1*T_0_0 + X_0_2*T_0_1 + X_1_1*T_1_0 + X_1_2*T_1_1 + b = 26
+  // O_1_0 = X_1_0*T_0_0 + X_1_1*T_0_1 + X_2_0*T_1_0 + X_2_1*T_1_1 + b = 38
+  // O_1_1 = X_1_1*T_0_0 + X_1_2*T_0_1 + X_2_1*T_1_0 + X_2_2*T_1_1 + b = 44
+  // (b is ignored).
+  //
+  // == Table for derivates ==
+  //         d|X00|X01|X02|X10|X11|X12|X20|X21|X22
```
Comment (SplitInfinity) [inline tests/unittests/OperatorGradTest.cpp]: Similarly, consider writing the terms in this table as `dO_{0,0}`, `d|X_{0,0}|...`, `T_{0,0}`, etc.

File: tests/unittests/OperatorGradTest.cpp [hunk 3]
```diff
@@ -105,6 +105,57 @@ TEST_P(OperatorGradTest, concat) {
...
+  // ----------------------------------------------
+  // dO_0_0   |T00|T01| 0 |T10|T11| 0 | 0 | 0 | 0
+  // dO_0_1   | 0 |T00|T01| 0 |T10|T11| 0 | 0 | 0
+  // dO_1_0   | 0 | 0 | 0 |T00|T01| 0 |T10|T11| 0
+  // dO_1_1   | 0 | 0 | 0 | 0 |T00|T01| 0 |T10|T11
+  //
+  // Based on the table above, dF/dX can be easily computed by
+  // dF/dX_i_j = sum(dF/dO_k_l * dO_k_l/dX_i_j) for all k,l.
```
Comment (SplitInfinity) [inline tests/unittests/OperatorGradTest.cpp]: Same here: `dF/dX_{i,j} = sum((dF/dO_{k_l}) * (dO_{k,l}/dX_{i,j})) for all k, l`

File: tests/unittests/OperatorGradTest.cpp [hunk 4]
```diff
@@ -105,6 +105,57 @@ TEST_P(OperatorGradTest, concat) {
...
+      false);
+  bindings_.allocate(X)->getHandle() = {0, 1, 2, 3, 4, 5, 6, 7, 8};
+
+  ConvolutionNode *O =
+      F_->createConv(bindings_, "conv", X, 1 /* outChannels */, 2 /* kernel */,
+                     1 /* stride */, 0 /* pad */, 1 /* group */);
+
+  // Note that filter is trainable, but probably it's okay since we run the
+  // graph only once.
+  bindings_.get(llvm::cast<Placeholder>(O->getFilter()))->getHandle() = {0, 1,
```
Comment (nadavrot) [inline tests/unittests/OperatorGradTest.cpp]: This statement is already two lines. Let's split it in a more readable way. Like:

auto *filter = O->getFilter();
bindings_.get(llvm::cast<Placeholder>(filter))->getHandle() = {0, 1, 2, 3};

File: tests/unittests/OperatorGradTest.cpp [hunk 5]
```diff
@@ -105,6 +105,57 @@ TEST_P(OperatorGradTest, concat) {
...
+      ElemKind::FloatTy, {1, 3 /* width */, 3 /* height */, 1 /* depth */}, "x",
+      false);
+  bindings_.allocate(X)->getHandle() = {0, 1, 2, 3, 4, 5, 6, 7, 8};
+
+  ConvolutionNode *O =
+      F_->createConv(bindings_, "conv", X, 1 /* outChannels */, 2 /* kernel */,
+                     1 /* stride */, 0 /* pad */, 1 /* group */);
+
+  // Note that filter is trainable, but probably it's okay since we run the
+  // graph only once.
```
Comment (nadavrot) [inline tests/unittests/OperatorGradTest.cpp]: This can be:  llvm::cast<Placeholder> -> cast<Placeholder>.

Add using llvm::cast at the top of the file.

**Comment by SplitInfinity** [comment]
I would recommending trying to create a template function out of this so that we can create several tests with different padding, strides, and kernel sizes.

**Pull request closed without merging**